numpy/core/src/multiarray/dtype_transfer.c

```diff
@@ -2638,286 +2640,104 @@ get_fields_transfer_function(int aligned,
         data->base.clone = &_field_transfer_data_clone;
         fields = &data->fields;
 
-        key = PyTuple_GET_ITEM(names, 0);
+        key = PyTuple_GET_ITEM(src_names, 0);
         tup = PyDict_GetItem(src_dtype->fields, key);
-        if (!PyArg_ParseTuple(tup, "Oi|O", &src_fld_dtype,
-                                                &src_offset, &title)) {
-            PyArray_free(data);
+        if (!PyArg_ParseTuple(tup, "Oi|O",
+                              &src_fld_dtype, &src_offset, &title)) {
             return NPY_FAIL;
         }
-        field_count = 0;
-        /*
-         * Special case bool type, the existence of fields implies True
```

Comment:
Note discussion of removal of this section [here](https://github.com/numpy/numpy/pull/6053#discussion_r123046754).